tests/test_timer.cpp

```diff
@@ -32,164 +33,555 @@ using namespace testing;
 
 namespace
 {
-struct Timer : public testing::Test
+struct MockTimerSyncFuncs : public mpu::TimerSyncFuncs
 {
-    std::atomic_bool timedout = false;
+    using TimerSyncFuncs::TimerSyncFuncs;
+
+    MOCK_METHOD1(notify_all, void(std::condition_variable&));
+    MOCK_METHOD2(wait, void(std::condition_variable&, std::unique_lock<std::mutex>&));
+    MOCK_METHOD3(wait_for, std::cv_status(std::condition_variable&, std::unique_lock<std::mutex>&,
+                                          const std::chrono::duration<int, std::milli>&));
+
+    MP_MOCK_SINGLETON_BOILERPLATE(MockTimerSyncFuncs, TimerSyncFuncs);
+};
+
+struct TestTimer : public testing::Test
+{
+    std::atomic_bool timedout{false}, test_notify_called{false}, timer_running{false}, timer_paused{false};
     std::condition_variable cv;
     std::mutex cv_m;
+    const std::chrono::milliseconds default_timeout{10};
+    const std::chrono::seconds default_wait{1};
+
+    MockTimerSyncFuncs::GuardedMock attr{MockTimerSyncFuncs::inject()};
+    MockTimerSyncFuncs* mock_timer_sync_funcs = attr.first;
 };
 } // namespace
 
-TEST_F(Timer, times_out)
+TEST_F(TestTimer, times_out)
 {
-    mpu::Timer t{duration_cast<seconds>(1ms), [this]() {
+    EXPECT_CALL(*mock_timer_sync_funcs, wait_for(_, _, _)).WillOnce([this](auto&, auto& timer_lock, auto&) {
+        std::unique_lock<std::mutex> test_lock(cv_m);
+
+        // Unlock the Timer's internal lock
+        timer_lock.unlock();
+
+        // Wait for notification from main test thread
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
+
+        return std::cv_status::timeout;
+    });
+
+    // notify_all is called twice via the stop() in start() and the stop() in the dtor
+    EXPECT_CALL(*mock_timer_sync_funcs, notify_all(_)).Times(2).WillRepeatedly(Return());
```

Comment:
I don't think it's needed. If no action is specified for a particular mock call, it will just do the default. In void-returning functions, that is just no-op.

FWIW, in functions returning another type, the mock will try to default construct an object of that type. I think it throws if there is no default constructor for the return type, so an action must be provided in that case.